tensorflow/contrib/rnn/python/ops/rnn_cell.py

```diff
@@ -661,6 +663,109 @@ def _make_tf_features(self, input_feat):
+    of a matrix the echo state property necessary condition `max(eig(W)) < 1` is approximated
+    scaling the norm 2 of the reservoir matrix which is an upper bound of the spectral radius.
+    See https://en.wikipedia.org/wiki/Matrix_norm#Induced_norm
+
+  """
+
+  def __init__(self, num_units, wr2_scale=0.7, connectivity=0.1, leaky=1.0, activation=math_ops.tanh,
+               win_init=init_ops.random_normal_initializer(),
+               wr_init=init_ops.random_normal_initializer(),
+               bias_init=init_ops.random_normal_initializer()):
+    """Initialize the Echo State Network Cell.
+
+    Args:
+      num_units: Int or 0-D Int Tensor, the number of units in the reservoir
+      wr2_scale: desired norm2 of reservoir weight matrix.
+        `wr2_scale < 1` is a sufficient condition for echo state property.
+      connectivity: connection probability between two reservoir units
+      leaky: leaky parameter
+      activation: activation function
+      win_init: initializer for input weights
+      wr_init: used to initialize reservoir weights before applying connectivity mask and scaling
+      bias_init: initializer for biases
+    """
+    self._num_units = num_units
+    self._leaky = leaky
+    self._activation = activation
+
+    def _wr_initializer(shape, dtype):
+      wr = wr_init(shape, dtype=dtype)
+
+      connectivity_mask = \
+        math_ops.cast(
+          math_ops.less_equal(
+            random_ops.random_uniform(shape),
+            connectivity),
+          dtype)
+
+      wr = math_ops.mul(wr, connectivity_mask)
+
+      wr_norm2 = math_ops.sqrt(
+        math_ops.reduce_sum(
+          math_ops.square(wr)))
+
+      is_norm_0 = math_ops.cast(math_ops.equal(wr_norm2, 0), dtype)
+
+      wr = wr * wr2_scale / (wr_norm2 + 1 * is_norm_0)
+
+      return wr
+
+    self._win_initializer = win_init
+    self._bias_initializer = bias_init
+    self._wr_initializer = _wr_initializer
+
+  @property
+  def output_size(self):
+    return self._num_units
+
+  @property
+  def state_size(self):
+    return self._num_units
+
+  def __call__(self, inputs, state, scope=None):
+    """ Run one step of ESN Cell
+
+        Args:
+          inputs: `2-D Tensor` with shape `[batch_size x input_size]`.
+          state: `2-D Tensor` with shape `[batch_size x self.state_size]`.
+          scope: VariableScope for the created subgraph; defaults to class `ESNCell`.
+
+        Returns:
+          A tuple `(output, new_state)`, computed as
+          `output = new_state = (1 - leaky) * state + leaky * activation(Win * input + Wr * state + B)`.
+
+        Raises:
+          ValueError: if `inputs` or `state` tensor size mismatch the previously provided dimension.
+          """
+
+    input_size = inputs.get_shape().as_list()[1]
+    dtype = inputs.dtype
+
+    with vs.variable_scope(scope or type(self).__name__):  # "ESNCell"
+
+      win = vs.get_variable("InputMatrix", [input_size, self._num_units], dtype=dtype,
```

Comment:
it is often leads to faster computation e.g. on GPUs to create a single matrix, e.g., InputMatrix equivalent to concatenating ReservoirMatrix, perform a single matmul, and then split the results.

In fact, here you should probably just use contrib.layers.linear instead of calling the matmul + get_variable yourself.